### PR TITLE
[OPERATOR-425] storage node label validation

### PR DESCRIPTION
**Pull request by haibinxie**
**What this PR does / why we need it**:

**Which issue(s) this PR fixes** (optional)
Closes #

**Special notes for your reviewer**:

### CODE REVIEW

File: pkg/controller/storagecluster/storagecluster.go
Comment (piyush-nimbalkar) [inline pkg/controller/storagecluster/storagecluster.go:395]: * to have __only__ 1 key
This makes it clear that only 1 key is allowed

File: pkg/controller/storagecluster/storagecluster.go
Comment (piyush-nimbalkar) [inline pkg/controller/storagecluster/storagecluster.go:403]: nit: We could just check `key == ""` instead of the `first` boolean, right?

File: pkg/controller/storagecluster/storagecluster.go
Comment (piyush-nimbalkar) [inline pkg/controller/storagecluster/storagecluster.go:375]: Shouldn't `key` be the expected and `Spec.CloudStorage.NodePoolLabel` be the actual?

File: pkg/controller/storagecluster/storagecluster.go
Comment (piyush-nimbalkar) [inline pkg/controller/storagecluster/storagecluster.go:350]: If cloud storage is not empty, then we should not even allow `nodeName` selector. The problem would be that if different nodes define different cloud storage configurations, then we won't be able to honor them on restarts. Same as different label selector key problem we are trying to solve/validate.

File: pkg/controller/storagecluster/controller_test.go
Comment (piyush-nimbalkar) [inline pkg/controller/storagecluster/controller_test.go:507]: Let's add a case where one of the nodes section does not have cloud storage.

**Review by piyush-nimbalkar — COMMENTED** [review]
Overall looks good. 
Just need to handle the case of node name selector.

**Comment by piyush-nimbalkar** [comment]
Take a look at the travis build as well. I think formatting is broken in few places.

File: pkg/controller/storagecluster/storagecluster.go
Comment (haibinxie) [reply pkg/controller/storagecluster/storagecluster.go:403]: correct

File: pkg/controller/storagecluster/storagecluster.go
Comment (haibinxie) [reply pkg/controller/storagecluster/storagecluster.go:375]: if nodePoolLabel is configured, then it should be used and expected. well depends on how we understand it, either one can be source of truth

File: pkg/controller/storagecluster/storagecluster.go
Comment (haibinxie) [reply pkg/controller/storagecluster/storagecluster.go:350]: "different nodes define different cloud storage configurations, then we won't be able to honor them on restarts", the node name does not change upon restarts and the cloud storage configuration should be honored right?

Comment (piyush-nimbalkar) [reply pkg/controller/storagecluster/storagecluster.go:350]: The issue here is how clouddrive is implemented. Currently the entire logic is in the px layer. The operator sends the params to the px container, but when failover of clouddrive happens in the background, the original spec may not be honored. This could be confusing to the user. That is why we are adding these validations. We can talk offline if you need more context.

File: pkg/controller/storagecluster/storagecluster.go
Comment (piyush-nimbalkar) [reply pkg/controller/storagecluster/storagecluster.go:375]: hmm.. that is true
Still I feel we should call `key` as expected so it tells the user that the `cloudStorage.NodePoolLabel` is the user defined key and it should match the expected value in `key`. Users can only modify that and not the 'expected'. Wdyt?

**Comment by haibinxie** [comment]
@piyush-nimbalkar I have resolved the comment, this may cause existing storagecluster to fail validation after upgrading operator, this could be bad to users, and it could be very tricky to update storagecluster if users run into this. let's discuss the impact before merging.

File: pkg/controller/storagecluster/storagecluster.go
Comment (piyush-nimbalkar) [inline pkg/controller/storagecluster/storagecluster.go:385]: So, in case of nodeName selector, we are going to not enforce the validation?

Comment (haibinxie) [reply pkg/controller/storagecluster/storagecluster.go:385]: i will add it back later. thinking of only enforcing the validation on fresh installation, the problem is the first time we know status is empty it is fresh installation we can fail the validation then update the status in storagecluster, but in second reconcile loop the status is updated so we no longer know it is fresh installation.

Comment (haibinxie) [reply pkg/controller/storagecluster/storagecluster.go:400]: resolved. also add check to skip validation on existing cluster.

**Comment by haibinxie** [comment]
@piyush-nimbalkar please take a final look, i added a change to skip validation is storage pod exists.

**Review by piyush-nimbalkar — APPROVED** [review]
Can you add tests for the two cases you added? 
- node name used with cloudstorage
- storage pods present and see that the validation is skipped

**Comment by haibinxie** [comment]
> Can you add tests for the two cases you added?
> 
> * node name used with cloudstorage
> * storage pods present and see that the validation is skipped

@piyush-nimbalkar added in a new commit. will merge once the build passes.

**Pull request merged**
